tests/test_core_network.py

```diff
@@ -475,38 +475,63 @@ def test_attempt_to_simplify_already_simplified_network_throws_error():
+    assert not puma_network.is_simplified()
+    link_ids_pre_simplify = set(dict(puma_network.links()).keys())
 
-    assert n.is_simplified()
+    deleted_self_loops = puma_network.simplify()
+    assert puma_network.is_simplified()
 
-    link_ids_post_simplify = set(dict(n.links()).keys())
+    link_ids_post_simplify = set(dict(puma_network.links()).keys())
 
     assert link_ids_post_simplify & link_ids_pre_simplify
     new_links = link_ids_post_simplify - link_ids_pre_simplify
     deleted_links = link_ids_pre_simplify - link_ids_post_simplify
-    assert set(n.link_simplification_map.keys()) == deleted_links
-    assert set(n.link_simplification_map.values()) == new_links
-    assert (set(n.link_id_mapping.keys()) & new_links) == new_links
+    # check the links removed in simplification are all mapped to new links
+    assert set(puma_network.link_simplification_map.keys()) == deleted_links
+    # check even the removed loop links have the simplification mapping retained
+    assert set(puma_network.link_simplification_map.values()) == new_links | deleted_self_loops
+    # check all new links int he network have edge mappings
+    assert (set(puma_network.link_id_mapping.keys()) & new_links) == new_links
 
-    report = n.generate_validation_report()
+    report = puma_network.generate_validation_report()
 
     assert report['routing']['services_have_routes_in_the_graph']
     assert report['schedule']['schedule_level']['is_valid_schedule']
 
 
-def test_simplified_network_saves_to_correct_dtds(tmpdir, network_dtd, schedule_dtd):
-    n = read.read_matsim(path_to_network=puma_network_test_file, epsg='epsg:27700',
-                         path_to_schedule=puma_schedule_test_file)
+def test_simplify_does_not_oversimplify_PT_endpoints(puma_network):
+    assert not puma_network.is_simplified()
+    stops_at_risk = ['5221390681543854913', '5221390302070799085', '5221390323679791901']
+    for s in stops_at_risk:
+        assert puma_network.link(puma_network.schedule.stop(s).linkRefId)['length'] == 1
 
-    n.simplify()
+    deleted_self_loops = puma_network.simplify()
 
-    n.write_to_matsim(tmpdir)
+    for s in stops_at_risk:
+        assert puma_network.link(puma_network.schedule.stop(s).linkRefId)['length'] == 1
+
+
+def test_simplify_removes_loops_by_default_but_keeps_pt_stop_loops(puma_network):
+    deleted_self_loops = puma_network.simplify()
+    assert deleted_self_loops
+    for stop in puma_network.schedule.stops():
+        assert puma_network.link(stop.linkRefId)['from'] == puma_network.link(stop.linkRefId)['to']
+        assert puma_network.link(stop.linkRefId)['length'] == 1
+
+
+def test_simplified_network_saves_to_correct_dtds(tmpdir, network_dtd, schedule_dtd, puma_network):
+    try:
+        puma_network.simplify()
+    except Exception as e:
+        raise RuntimeError(f"Error simplifying network: {e}, check other simplification tests")
```

Comment:
I wasn't thinking about `simplify` possibly throwing an exception - if it does that, it will be clear what caused the test to fail, without wrapping and rethrowing the exception. I don't think rethrowing the exception adds anything.

What I was driving at is that we're calling `simplify` - what if it doesn't throw an exception, **but** it's broken in some way? Testing around that and that alone would have a network, simplify it, then make assertions about it (without serialising it before making those assertions). I think we have tests like that, right?

But if we also write the network and it doesn't validate against the DTD, is that because `simplify` has a bug, or because `write_to_matsim` has a bug? Or both? In an ideal world, we test `write` by saying "given a known network, when we serialise it to disk, it should blah, blah (validate against the DTD, etc.)". Whether or not the network was simplified first shouldn't matter, we're just testing that a network serialises properly.

It seems like the thing we're testing here is that a simplified network serialises correctly, so is there a way to have a simplified network with which to test that **without** calling simplify? If that's possible, it would be a better way to test the serialisation of the simplified network. As it stands, this is kind of a pipeline of two operations with some assertions at the end of the pipeline.